Sofa/Component/Constraint/Lagrangian/Correction/src/sofa/component/constraint/lagrangian/correction/UncoupledConstraintCorrection.cpp

```diff
@@ -160,7 +161,7 @@ int UncoupledConstraintCorrectionClass = core::RegisterObject("Component computi
         .add< UncoupledConstraintCorrection< Vec2Types > >()
         .add< UncoupledConstraintCorrection< Vec3Types > >()
         .add< UncoupledConstraintCorrection< Rigid3Types > >()
-    ;
+        .setTemplateDeductionMethod(sofa::core::getTemplateFromMechanicalState);
```

Comment:
The "problem" I see when the template deduction is set in the registry, is for the SofaCUDA plugin, where the component is added in the registry a second time with different template parameters. If I understood correctly, the components from SofaCUDA will not "inherit" from the same deduction rule, creating an inconsistency. So it must be added again in SofaCUDA, but it is easy to forget...

Reply:
I see two way of fixing that:
- clean the object factory implementation so there is one deduction rule per component registered (which actually what it should be).  There exists several previous experimental PR on how we could refactor the object factory but this is far beyond the template deduction system.  And thus not recommending coupling these two topics. 

- only use the static method way of defining a template deduction rule. It make this PR simpler :)

Reply:
I know you're not fan of the static method. Could you give the reasons?

Reply:
Because it is using the c++ name resolution order to implement static method overrides. 

This make the static method override very implicit and also very fragile. 
Implicit because: when looking at a static method signature in a code base, nothing tells the developer that the method is not just a normal "individual" but is in fact "part" of a complex multi-class design involving at least the parent and child classes as well as a third party object (here the ObjectFactory) will call the static method.  
Fragile: because if you decide to refactor your design, and thus removing the static method in the class one and its corresponding call (in the ObjectFactory in our case), the compiler will not tell you anything that all the intermediates method are useless.

So to me it would way better if there was a c++ keyword to specify that a static method can be "override" (a kind of static overridable) and a second keyword to indicate a static method actually "override". 
So something similar as what we have with the "virtual/override" keyword.

Reply:
This is the pattern:
```cpp
#include<string>
#include <iostream>

class ParentClass
{
public:
    /// This method is used in ObjectFactoryPrintYourName... please don't remove it ! 
    static std::string GetName(){ return "BaseClass"; }
};

class ChildClass : public ParentClass
{
public:
     /// This method overrides the ParentClass::GetName()...
    static std::string GetName(){  return "ChildClass"; }
};

class GrandChildClass : public ChildClass
{
public:
     /// This method overrides the ChildClass::GetName()...
    static std::string GetName(){  return "GrandChildClass"; }
};


template<class T>
void ObjectFactoryPrintYourName()
{
    std::cout << T::GetName() << std::endl;
}

int main(int argc, char** argv)
{
    ObjectFactoryPrintYourName<ParentClass>();
    ObjectFactoryPrintYourName<ChildClass>();
    ObjectFactoryPrintYourName<GrandChildClass>();
}
```
Could we make better than that ?

Reply:
After thinking a bit more on that... I'm pretty sure that the right solution to the problem pointed by @alxbilger would not exists if there was a properly implemented ObjectFactory where there is a single deductionRule for a className. 
I don't see any clear added value in implementing a system that allows FEMForceField<CudaVec3> to deduced its type i a different way the  FemForceField<Vec3> do.